File: Data Structures & Algorithms/eating-bananas/submission-4.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        Arrays.sort(piles);
        int l = 1;
        int r = Arrays.stream(piles).max().getAsInt();
        int k = 0;
        int min_k = Integer.MAX_VALUE;
        while(l <= r){
            k = (l+r)/2;
            int time = 0;
            for(int i = 0; i < piles.length; ++i){
                time += Math.ceil((double)piles[i]/k);
            }
            if(time <= h){
                r = k -1;
                min_k = Math.min(min_k, k);
            }else{
                l = k+1;
            }

        }
        return min_k;
    }
}

File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-4.py
class Solution:
    def minCostClimbingStairs(self, cost: list[int]) -> int:
        n  = len(cost)
        dp = [0] * (n+1)
        dp[0] = 0
        dp[1] = 0
        dp[2] = min(cost[0], cost[1])
        print(dp[2])

        for i in range(3, n + 1):
            dp[i] = min(
                dp[i-1] + cost[i-1],
                dp[i-2] + cost[i-2]
            )
        
        return dp[n]

